### ISSUE TITLE
[bug] stack-trace when installing packages with lockfiles with --lockfile-packages

### ISSUE THREAD

**Issue by Wiral** [conan-io/conan#13972]
I'm using conan 2.0.5 (ubuntu 20), artifactory Enterprise Plus 7.55.6 rev, and get an error when trying to pull package using lockfile **locked down to package revision**.
The recipe is very simple and doesn't have any requires/dependencies. All it does is package some headers files and library for specified profile.
Detailed explanation:
* The package is present of the remote. As you can see, my recipe reference is `9ed4a2da81fa33ec7b4709278a3f7fba`.
```
conan list <conan_reference>#9ed4a2da81fa33ec7b4709278a3f7fba:*#* -r <remote> 
pps-conan-dev-sfo
  dvpe
    <conan_reference>
      revisions
9ed4a2da81fa33ec7b4709278a3f7fba (2023-05-18 12:02:22 UTC)
          packages
            2173785b22b02385e5b644db09ccbc9100463eb7
              revisions
                 a71640f81f4f23dddb8233d1e7b861b8 (2023-05-18 12:02:25 UTC)
                   info
                    settings
                      arch: x86_64
                      compiler: gcc
                      compiler.cppstd: 11
                      compiler.libcxx: libstdc++
                      compiler.version: 4.8
                      os: Linux
```
* Lockfile created based on the package above:
```
conan lock create --lockfile-packages --requires <conan_reference> --lockfile-out l.lock -pr:b linux_x86_64_gcc_4_8 -pr:h linux_x86_64_gcc_4_8 -r <remote>
```
This results in the following lockfile:
```
{
    "version": "0.5",
    "requires": [
        [
            "<conan_reference>#9ed4a2da81fa33ec7b4709278a3f7fba%1684411342.717",
            {
                "2173785b22b02385e5b644db09ccbc9100463eb7": "a71640f81f4f23dddb8233d1e7b861b8"
            }
        ]
    ],
    "build_requires": [],
    "python_requires": []
```
As you can see, recipe, package_id and revision match the package present on the remote.
* install the package using the genereated lockfile
```
conan install --requires <conan_reference> -l l.lock -r <remote> -b never -pr:b linux_x86_64_gcc_4_8 -pr:h linux_x86_64_gcc_4_8
Using lockfile: '/home/jsobo/l.lock'

 

======== Input profiles ========
Profile host:
[settings]
arch=x86_64
compiler=gcc
compiler.cppstd=11
compiler.libcxx=libstdc++
compiler.version=4.8
os=Linux

 

Profile build:
[settings]
arch=x86_64
compiler=gcc
compiler.cppstd=11
compiler.libcxx=libstdc++
compiler.version=4.8
os=Linux

 


======== Computing dependency graph ========
Graph root
    cli
Requirements
    <conan_reference>#9ed4a2da81fa33ec7b4709278a3f7fba - Cache

 

======== Computing necessary packages ========
Requirements
    <conan_reference>#9ed4a2da81fa33ec7b4709278a3f7fba:2173785b22b02385e5b644db09ccbc9100463eb7#a71640f81f4f23dddb8233d1e7b861b8 - None

 

======== Installing packages ========
Traceback (most recent call last):
  File "/home/jsobo/.local/lib/python3.8/site-packages/conan/cli/cli.py", line 272, in main
    cli.run(args)
  File "/home/jsobo/.local/lib/python3.8/site-packages/conan/cli/cli.py", line 171, in run
    command.run(self._conan_api, self._commands[command_argument].parser, args[0][1:])
  File "/home/jsobo/.local/lib/python3.8/site-packages/conan/cli/command.py", line 157, in run
    info = self._method(conan_api, parser, *args)
  File "/home/jsobo/.local/lib/python3.8/site-packages/conan/cli/commands/install.py", line 82, in install
    conan_api.install.install_binaries(deps_graph=deps_graph, remotes=remotes)
  File "/home/jsobo/.local/lib/python3.8/site-packages/conan/api/subapi/install.py", line 22, in install_binaries
    installer.install(deps_graph, remotes)
  File "/home/jsobo/.local/lib/python3.8/site-packages/conans/client/installer.py", line 252, in install
    self._handle_package(package, install_reference, None, handled_count, package_count)
  File "/home/jsobo/.local/lib/python3.8/site-packages/conans/client/installer.py", line 298, in _handle_package
    assert package.binary in (BINARY_CACHE, BINARY_BUILD, BINARY_DOWNLOAD, BINARY_UPDATE)
AssertionError

 

ERROR: 
```
Despite the expected recipe, package-id and package revision requested, the package is not found (None) which results in a stack trace and empty error message. It does not matter if i `conan remove '*' -f ` before doing the above steps. 
Everything works correctly when the lockfile is created without ` --lockfile-packages`. In that scenario latest available package is pulled, which is obviously the only one available.

**Comment by memsharded** [comment]
Hi @Wiral 

Thanks for your report. 
As a first note, I would like to highlight the following paragraph from the docs:

> The ``--lockfile-packages`` argument allows to create lockfiles that also lock down to the package revision, but 
it should not be necessary in the vast majority of cases, so it is discouraged in the general case.

This feature is really not documented yet, beyond the command line arguments that appear in ``conan xxx -h``. We might even not document it fully, or document it with strong caveats or warnings, still not decided.

Please take into account that needing to lock package revisions is a symptom that the process is not well defined in the first place. As a general rule, only 1 package revision of any given package_id should exist. If you got more than 1 package revision that is an indicator something went wrong. Having to rely on lockfiles that lock individual package revisions is an attempt to patch the issue, but without really addressing the core issue.

Please let me know if you have any question regarding this.

Besides that, it is true that Conan commands should not give stack trace, and that should always be considered an error to be fixed, so we will have a look when possible.

Thanks very much again for your report!

**Comment by Wiral** [comment]
I'm aware of what you wrote regarding one package revision, but I'm trying to prevent any "silent" package changes happening in my project, in case some of our dependencies happens to have ill-defined release process.

I was expecting the lock mechanism to somewhat loosely reflect conan1.X behavior when locking onto package revision. My exemplary use-case seems to be one of the most basic use-cases for this `--lockfile-packages` flag I can think of, which now completely fails and therefore I don't see this flag being useful at all.

Since it's not documented, I suspect you can't tell me what's the use-case for this switch? Should I treat it as non-tested/not-implemented at this state?

**Comment by memsharded** [comment]
> I was expecting the lock mechanism to somewhat loosely reflect conan1.X behavior when locking onto package revision.

Conan 1.X were overly complicated and fragile, and still not able to implement some important and necessary functionality to implement CI flows. Having them modeling package revisions by default was one of the sources of these problems. Also related to ``package_revision_mode``, which has been removed in Conan 2.0, because of exactly the same reason. It is not possible to parameterize the ``package_ids`` based on the package-revisions of the dependencies, because dependencies shouldn't have more than 1 package-revision. So lockfiles in Conan 2.0 are a complete new solution, from scratch.


> Since it's not documented, I suspect you can't tell me what's the use-case for this switch? Should I treat it as non-tested/not-implemented at this state?

Yes, that would be the best. We provided some internal implementation just to make sure the core modeling, serialization and management of lockfiles files makes this possible in the future, so the current state doesn't become a future blocker. But as commented, it shouldn't be considered as a solution for ill-defined release processes, but the solution should be done upstream in those packages processes.
